File: Python/ByTier/E/2022_E2_Billetes_MX_Hard_Version.py
# Problem: CF 2022 E2 - Billetes MX (Hard Version)
# https://codeforces.com/contest/2022/problem/E2

"""
E2. Billetes MX (Hard Version)

Algorithm:
- This code solves a problem involving assigning values to cells of a grid such that all subgrids maintain a specific XOR property.
- Uses Union-Find (Disjoint Set Union) with path compression and distance tracking via XOR operations to model constraints between cells.
- Time Complexity: O((n+m+k+q)*α(n+m)) where α is the inverse Ackermann function, effectively constant for practical purposes.
- Space Complexity: O(n + m)

Techniques:
- Union-Find with distance tracking using XOR (for maintaining consistency of XOR conditions).
- Modular exponentiation to compute powers of 2 modulo 10^9 + 7.

Explanation:
Each cell in the grid is treated as a node, and the constraint that four corner values of any subgrid must XOR to zero can be modeled via a Union-Find structure.
The idea is to relate nodes by using XOR distances, ensuring that constraints are satisfied when units are merged.
"""

import io
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def get_root(s):
    w = 0
    while s ^ root[s]:  # Find root with path compression and update weight
        s, w = root[s], w ^ dist[s]
    return s, w


def unite(s, t, w):
    s, ws = get_root(s)
    t, wt = get_root(t)
    if s == t:
        return 1 if not ws ^ wt ^ w else -1  # Return 1 if conflict (inconsistent), -1 otherwise
    if rank[s] < rank[t]:
        s, t = t, s
    if rank[s] == rank[t]:
        rank[s] += 1
    root[t], dist[t] = s, dist[s] ^ ws ^ wt ^ w  # Update parent and distance to root
    return 0


t = int(input())
ans = []
mod = pow(10, 9) + 7
l = 2 * pow(10, 5) + 5
p = [1] * l
for i in range(1, l):
    p[i] = (p[i - 1] << 30) % mod  # Precompute powers of 2^30 mod (10^9 + 7)
for _ in range(t):
    n, m, k, q = map(int, input().split())
    l = n + m
    root = [i for i in range(l + 1)]
    rank = [1 for _ in range(l + 1)]
    dist = [0] * (l + 1)
    ok, u = 1, l - 1  # ok tracks validity of solution; u = number of free variables
    for _ in range(k):
        r, c, v = map(int, input().split())
        if ok:
            f = unite(r, c + n, v)  # Unite row and column node with value v
            if f == -1:
                ok = 0
            u -= f ^ 1  # Decrease free variables if not conflict
    ans0 = [0] * (q + 1)
    if ok:
        ans0[0] = p[u]  # Initial answer based on number of independent variables
    for i in range(1, q + 1):
        r, c, v = map(int, input().split())
        if ok:
            f = unite(r, c + n, v)  # Update with new value at (r,c)
            if f == -1:
                ok = 0
            u -= f ^ 1  # Adjust free variables count
        if ok:
            ans0[i] = p[u]  # Final answer for query state
    ans.append("\n".join(map(str, ans0)))
sys.stdout.write("\n".join(ans))


# https://github.com/VaHiX/codeForces/

File: CPP/ByRound/2022/2022_E2_Billetes_MX_Hard_Version.cpp
// Problem: CF 2022 E2 - Billetes MX (Hard Version)
// https://codeforces.com/contest/2022/problem/E2

/*
 * Problem: E2. Billetes MX (Hard Version)
 *
 * Purpose:
 *   This code solves a problem involving assigning values to a grid such that
 *   all subgrids satisfy a XOR constraint on their corners. It uses Union-Find
 *   with path compression and value tracking to manage connected components and
 *   validate XOR constraints. After each query, it updates the component count
 *   and returns the number of valid assignments based on the remaining free
 *   variables, calculated as 2^(number of connected components - fixed cells).
 *
 * Algorithms:
 *   - Union-Find (Disjoint Set Union) with path compression and value tracking
 *   - XOR constraint propagation in a 2D grid using coordinate mapping
 *
 * Time Complexity:
 *   O((n + m + k + q) * α(n + m)) where α is the inverse Ackermann function.
 *   Each DSU operation (union/find) takes nearly constant time.
 *
 * Space Complexity:
 *   O(n + m) for the DSU structures and arrays.
 */

#include <iostream>
#include <utility>
using namespace std;

const int N = 2e5 + 5, M = 1e9 + 7;
int t, n, m, k, q, p[N], x, y, z, dsu[N], sz[N], val[N], ok, cmps;

// Merge two sets with XOR constraint c (x ^ y = c)
void mrg(int a, int b, int c) {
  // Traverse to root while propagating XOR values
  while (a != dsu[a])
    c ^= val[a], a = dsu[a];
  while (b != dsu[b])
    c ^= val[b], b = dsu[b];

  // If already in same component, check XOR consistency
  if (a == b) {
    if (c)
      ok = 0;  // Inconsistent constraint
    return;
  }

  // Union by size: make sure 'a' is smaller
  if (sz[a] > sz[b])
    swap(a, b);

  // Merge a into b and update val and size
  dsu[a] = b, sz[b] += sz[a], val[a] = c, cmps--;
}

int main() {
  ios::sync_with_stdio(0), cout.tie(0), cin.tie(0);
  
  // Precompute powers of 2^30 mod M for fast calculation
  for (p[0] = x = 1; x < N; x++)
    p[x] = p[x - 1] * (1LL << 30) % M;

  cin >> t;
  while (t--) {
    cin >> n >> m >> k >> q;
    
    ok = 1, cmps = n + m - 1;  // Initially, there are n+m-1 independent components
    
    // Initialize DSU structures
    for (x = 1; x <= n + m; x++)
      dsu[x] = x, sz[x] = 1, val[x] = 0;

    // Process fixed cells and build constraints
    while (k--)
      cin >> x >> y >> z, mrg(x, y + n, z);  // Map column to n+y to distinguish rows from columns

    cout << ok * p[cmps] << "\n";

    // Process queries and output results after each update
    while (q--) {
      cin >> x >> y >> z, mrg(x, y + n, z), cout << ok * p[cmps] << "\n";
    }
  }
}


// https://github.com/VaHiX/codeForces/